Typecheck `else` expression of `let-else` to ensure it diverges

CohenArthur opened the pull request:
The `else` expression of a `let-else` must always diverge, which we enforce during typecheck. Needs #3468

CohenArthur committed:
tychk: Ensure diverging else of let-else is actually diverging

Make sure the `else` expression resolves to the never type.

gcc/rust/ChangeLog:

	* typecheck/rust-hir-type-check-stmt.cc (TypeCheckStmt::visit): Check diverging else
	against `!`.

gcc/testsuite/ChangeLog:

	* rust/compile/let-else-invalid-type.rs: New test.

CohenArthur commented on gcc/rust/typecheck/rust-hir-type-check-stmt.cc:
```diff
@@ -97,6 +100,22 @@ TypeCheckStmt::visit (HIR::LetStmt &stmt)
+	  return;
+	}
+
+      // FIXME: Is that enough? Do we need to do something like
+      // `append_reference` here as well?
```
@philberty I'm not sure if there's anything else we need to do here. I was trying to use something like `coercion_site(...)` but obviously it doesn't work because any type is coercible to `!`, which is not the behavior we actually want here

powerboat9 replied:
`!` should be coercible to any other type, but not the other way around

powerboat9 replied:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=83c424a3b36b5c35edcf6d8995c97ee2

It looks like an `if` and `else` branch with the same types should also work.

philberty reviewed (approved):
LGTM this is enough

CohenArthur committed:
nr: Fix typo when resolving let-else expressions

gcc/rust/ChangeLog:

	* resolve/rust-late-name-resolver-2.0.cc (Late::visit): Visit the `else` expression instead
	of the `init` one if the let is a let-else.
